### PR TITLE
Use `ControlFlow` in visitors.

### PR DESCRIPTION
**Pull request by Jarcho**
Follow up to #121256 

This does have a few small behaviour changes in some diagnostic output where the visitor will now find the first match rather than the last match. The change in `find_anon_types.rs` has the only affected test. I don't see this being an issue as the last occurrence isn't any better of a choice than the first.

### CODE REVIEW

File: compiler/rustc_infer/src/infer/error_reporting/nice_region_error/find_anon_type.rs [hunk 1]
```diff
@@ -145,23 +137,30 @@ impl<'tcx> Visitor<'tcx> for FindNestedTypeVisitor<'tcx> {
             }
             // Checks if it is of type `hir::TyKind::Path` which corresponds to a struct.
             hir::TyKind::Path(_) => {
-                let subvisitor = &mut TyPathVisitor {
-                    tcx: self.tcx,
-                    found_it: false,
-                    bound_region: self.bound_region,
-                    current_index: self.current_index,
+                // Prefer using the lifetime in type arguments rather than lifetime arguments.
+                intravisit::walk_ty(self, arg)?;
```
Comment (wesleywiser) [inline compiler/rustc_infer/src/infer/error_reporting/nice_region_error/find_anon_type.rs:141]: It looks like we have an extra call to `walk_ty` here? I don't see where this was happening in the old version of the code.

Comment (Jarcho) [reply compiler/rustc_infer/src/infer/error_reporting/nice_region_error/find_anon_type.rs:141]: That match arm used to continue to the `walk_ty` call at the end of the function.

**Comment by Jarcho** [comment]
Waiting on #121576 in case there's a conflict.

**Pull request merged**